pkg/csi/service/fcd/controller.go

```diff
@@ -224,15 +225,101 @@ func (c *controller) ControllerPublishVolume(
+		log.Errorf(msg)
+		return nil, status.Errorf(codes.Internal, msg)
+	} else if len(req.NodeId) == 0 {
+		msg := fmt.Sprintf("Node ID is a required parameter.")
+		log.Errorf(msg)
+		return nil, status.Errorf(codes.Internal, msg)
+	}
+
+	discoveryInfo, err := c.connMgr.WhichVCandDCByFCDId(ctx, req.VolumeId)
+	if err != nil {
+		msg := fmt.Sprintf("WhichVCandDCByFCDId(%s) failed. Err: %v", req.VolumeId, err)
+		log.Errorf(msg)
+		return nil, status.Errorf(codes.Internal, msg)
+	}
+
+	fcd := discoveryInfo.FCDInfo
+
+	vm, err := discoveryInfo.DataCenter.GetVMByDNSName(ctx, req.NodeId)
+	if err != nil {
+		log.Errorf("GetVMByDNSName(%s) failed. Err: %v", req.NodeId, err)
+		return nil, err
+	}
+
+	filePath := fcd.Config.Backing.(*types.BaseConfigInfoDiskFileBackingInfo).FilePath
+	options := &vclib.VolumeOptions{SCSIControllerType: vclib.PVSCSIControllerType}
+	diskUUID, err := vm.AttachDisk(ctx, filePath, options)
+	if err != nil {
+		log.Errorf("AttachDisk(%s = %s) failed. Err: %v", fcd.Config.Name, filePath, err)
+		return nil, err
+	}
+
+	log.Infof("AttachDisk(%s) succeeded with UUID: %s", filePath, diskUUID)
+
+	publishInfo := make(map[string]string, 0)
+	publishInfo[AttributeFirstClassDiskType] = FirstClassDiskTypeString
+	publishInfo[AttributeFirstClassDiskName] = fcd.Config.Name
+	publishInfo[AttributeFirstClassDiskParentType] = string(fcd.ParentType)
+	if fcd.ParentType == vclib.TypeDatastoreCluster {
+		publishInfo[AttributeFirstClassDiskParentName] = fcd.StoragePodInfo.Summary.Name
+		publishInfo[AttributeFirstClassDiskOwningDatastore] = fcd.DatastoreInfo.Info.Name
+	} else {
+		publishInfo[AttributeFirstClassDiskParentName] = fcd.DatastoreInfo.Info.Name
+	}
+
+	resp := &csi.ControllerPublishVolumeResponse{
+		PublishContext: publishInfo,
+	}
+
+	return resp, nil
 }
 
 func (c *controller) ControllerUnpublishVolume(
 	ctx context.Context,
 	req *csi.ControllerUnpublishVolumeRequest) (
 	*csi.ControllerUnpublishVolumeResponse, error) {
 
-	return nil, nil
+	//check for required parameters
+	if len(req.VolumeId) == 0 {
+		msg := fmt.Sprintf("Volume ID is a required parameter.")
+		log.Errorf(msg)
+		return nil, status.Errorf(codes.Internal, msg)
+	} else if len(req.NodeId) == 0 {
+		msg := fmt.Sprintf("Node ID is a required parameter.")
+		log.Errorf(msg)
+		return nil, status.Errorf(codes.Internal, msg)
+	}
+
+	discoveryInfo, err := c.connMgr.WhichVCandDCByFCDId(ctx, req.VolumeId)
+	if err != nil {
+		msg := fmt.Sprintf("WhichVCandDCByFCDId(%s) failed. Err: %v", req.VolumeId, err)
+		log.Errorf(msg)
+		return nil, status.Errorf(codes.Internal, msg)
+	}
+
+	fcd := discoveryInfo.FCDInfo
+
+	vm, err := discoveryInfo.DataCenter.GetVMByDNSName(ctx, req.NodeId)
+	if err != nil {
+		log.Errorf("GetVMByDNSName(%s) failed. Err: %v", req.NodeId, err)
+		return nil, err
+	}
+
+	filePath := fcd.Config.Backing.(*types.BaseConfigInfoDiskFileBackingInfo).FilePath
+	err = vm.DetachDisk(ctx, filePath)
```

Comment:
Can also use `SelectByBackingInfo` here to check if the disk is already detached.

Reply:
The idempotency takes place into DetachDisk:
```
	if device == nil {
		glog.Warningf("No virtual device found with diskPath: %q on VM: %q", vmDiskPath, vm.InventoryPath)
		return nil
	}
```
Returning nil is the disk is not attached.

Reply:
Ah yes, that's the first change in this PR, I had looked at DetachDisk in the current tree.